Tools/BiblioTech/Commands/CheckCidCommand.cs

```diff
@@ -32,7 +39,78 @@ protected override async Task Invoke(CommandContext context)
 
             var response = await checker.PerformCheck(cid);
             await Program.AdminChecker.SendInAdminChannel($"User {Mention(user)} used '/{Name}' for cid '{cid}'. Lookup-success: {response.Success}. Message: '{response.Message}' Error: '{response.Error}'");
+
+            if (response.Success)
+            {
+                if (cidStorage.IsCidUsed(cid))
+                {
+                    await context.Followup($"{response.Message}\n\nThis CID has already been used by another user. No role will be granted.");
+                    return;
+                }
+
+                if (cidStorage.AddCid(cid, user.Id))
+                {
+                    var guildUser = context.Command.User as IGuildUser;
+                    if (guildUser != null)
+                    {
+                        try
+                        {
+                            var role = context.Command.Guild.GetRole(ALTRUISTIC_ROLE_ID);
+                            if (role != null)
+                            {
+                                await guildUser.AddRoleAsync(role);
+                                await context.Followup($"{response.Message}\n\nCongratulations! You've been granted the Altruistic Mode role for checking a valid CID!");
+                                return;
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            await Program.AdminChecker.SendInAdminChannel($"Failed to grant Altruistic Mode role to user {Mention(user)}: {ex.Message}");
+                        }
+                    }
+                }
+            }
+
             await context.Followup(response.Message);
         }
     }
+
+    public class CidStorage
+    {
+        private readonly string filePath;
+        private static readonly object _lock = new object();
+
+        public CidStorage(string filePath)
+        {
+            this.filePath = filePath;
+            if (!File.Exists(filePath))
+            {
+                File.WriteAllText(filePath, string.Empty);
+            }
+        }
+
+        public bool AddCid(string cid, ulong userId)
+        {
+            lock (_lock)
+            {
+                var existingEntries = File.ReadAllLines(filePath);
+                if (existingEntries.Any(line => line.Split(',')[0] == cid))
+                {
+                    return false;
+                }
+
+                File.AppendAllText(filePath, $"{cid},{userId}{Environment.NewLine}");
```

Comment:
You can use AppendAllLines(...) then you don't have to worry about the newline character.